src/Type/Php/CallUserFuncArrayDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,71 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Arg;
+use PhpParser\Node\Expr\Array_;
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Expr\Variable;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\Constant\ConstantArrayType;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\NeverType;
+use PHPStan\Type\Type;
+use function count;
+
+class CallUserFuncArrayDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'call_user_func_array';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		$args = $functionCall->getArgs();
+		$defaultReturn = ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+
+		if (count($args) === 1 && $args[0]->unpack) {
+			$array = $scope->getType($args[0]->value);
+			if ($array instanceof ConstantArrayType) {
+				$count = $array->count();
+
+				if ($count instanceof ConstantIntegerType && $count->getValue() > 2) {
+					return new NeverType();
+				}
+			}
+
+			return $defaultReturn;
+		}
+
+		if (count($args) <= 1) {
+			return new NeverType();
+		}
+
+		$callbackType = $scope->getType($args[0]->value);
+		$callbackArgsType = $scope->getType($args[1]->value);
+
+		if ($callbackType->isCallable()->yes() && $callbackArgsType->isArray()->yes()) {
```

Comment:
What is this entire codeblock about? Why are there instanceof specific AST nodes like Array_ and Variable?

Reply:
this block is about unwrapping the passed in array-arg so the array-items can be re-wrapped into `Arg` objects and passed thru to the FuncCall

`call_user_func_array('CallUserFuncArray\fun3', [1 ,2 ,3])`